### PR TITLE
Adds support for querying a meeting time

### LINKED ISSUE #61: Create an ArrayList that initially has one TimeRange that is the entire day
- [x] `availableTimes` holds one TimeRange that is the entire day

@f-camp @billie-cat

### LINKED ISSUE #62: For each event analyze and adjust times in availableTimes
4 cases to take into account:
- [x] `eventWindow` doesn't overlap with the `currentAvailableTime` --> Add the time to `newTimes`
- [x] `eventWindow` fits entirely into `currentAvailableTime` --> Split the TimeRange into two and add to `newTimes`
- [x] `eventWindow` ends after `currentAvailableTime` starts --> push back `currentAvailableTime's` start until `eventWindow` ends and add to `newTimes`
- [x] `eventWindow` starts before `currentAvailableTime` ends --> move `currentAvailableTime's` end forward and add to `newTimes`

@billie-cat @f-camp

### PR DESCRIPTION
**Pull request by jaysaleh**
This pull request adds support for the `query` method in `FindMeetingQuery.java`. Method will find available time to schedule a meeting and return it in a `Collection<TimeRange>`.

### CODE REVIEW

**Review by f-camp — APPROVED** [review]

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 1]
```diff
@@ -14,10 +14,78 @@
 
 package com.google.sps;
 
+import java.util.*; 
 import java.util.Collection;
+import com.google.sps.TimeRange;
+import com.google.common.collect.Iterables;
 
+/** 
+ * Determines a list of available TimeRanges to schedule a meeting
```
Comment (billie-cat) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: I believe TimeRanges should be {@link TimeRanges} or something like that.

Comment (billie-cat) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: And same for MeetingRequest

Comment (billie-cat) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: tbh I would omit talking about these classes and just mention what this class does at a high level. For example:

Determines a list of available time slots to schedules a meeting given ...

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 2]
```diff
@@ -14,10 +14,78 @@
...
+ * given a MeetingRequest and list of events taking place that day. 
+ */
 public final class FindMeetingQuery {
+  
+  /** 
+   * Returns the open windows for which {@code request} can be scheduled in.
```
Comment (billie-cat) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: This comment doesn't talk about what 'events' is needed for. Can it be expanded?

**Review by billie-cat — CHANGES_REQUESTED** [review]

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 1 above)
Comment (jaysaleh) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: fixed

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 2 above)
Comment (jaysaleh) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: fixed

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 3]
```diff
@@ -14,10 +14,79 @@
 
 package com.google.sps;
 
+import java.util.*; 
 import java.util.Collection;
+import com.google.sps.TimeRange;
+import com.google.common.collect.Iterables;
 
+/** 
+ * Determines a list of available time slots to schedule a meeting given
```
Comment (billie-cat) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]:
```suggestion
 * Determines available time slots to schedule a meeting given
```

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 4]
```diff
@@ -14,10 +14,79 @@
...
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    Collection<String> requiredAttendees = request.getAttendees();
+    Collection<TimeRange> availableTimes = assembleTime(events, requiredAttendees, request.getDuration());
+    return adjustAvailableTimes;
+  }
+
+  /**
+   * Returns the open windows for which {@code request} can be scheduled in,
+   * after making {@code events} TimeRanges unavailable in {@code availableTimes}.
```
Comment (billie-cat) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]:
```suggestion
   * after making {@code events}' TimeRanges unavailable in {@code availableTimes}.
```

Comment (billie-cat) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: in your comment you should never reference internal code. So you should omit mention of availableTimes.

You can rephrase as: ... after making {@code events} TimeRanges unavailable for scheduling.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 5]
```diff
@@ -14,10 +14,79 @@
...
+    }
+
+    // Remove TimeRanges that are too short.
+    availableTimes.removeIf(item -> item.duration() < duration);
+
+    return availableTimes;
+  }
+
+  /**
+   * Returns {@code newTimes} which takes all TimeRanges in {@code availableTimes} and
```
Comment (billie-cat) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Here too, you should not reference internal variables. Please rephrase this comment.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 6]
```diff
@@ -14,10 +14,79 @@
...
+   * adjusts them to not include {@code eventWindow}.
+   */
+  public Collection<TimeRange> adjustAvailableTimes(Collection<TimeRange> availableTimes, TimeRange eventWindow) {
+    Collection<TimeRange> newTimes = new ArrayList<>();
+    for(TimeRange availableTime : availableTimes) {
+      // If timeTime doesn't overlap with eventWindow then availableTime is available.
```
Comment (billie-cat) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: These comments seem to be describing what the code is doing, so it seems the comments could be omitted.

**Review by billie-cat — CHANGES_REQUESTED** [review]

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 4 above)
Comment (jaysaleh) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Fixed

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 5 above)
Comment (jaysaleh) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Fixed

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 6 above)
Comment (jaysaleh) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Fixed

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 7]
```diff
@@ -14,10 +14,75 @@
...
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    Collection<String> requiredAttendees = request.getAttendees();
+    Collection<TimeRange> availableTimes = assembleTime(events, requiredAttendees, request.getDuration());
+    return adjustAvailableTimes;
+  }
+
+  /**
+   * Returns the open windows for which {@code request} can be scheduled in,
+   * after making {@code events}'' TimeRanges unavailable for scheduling.
```
Comment (billie-cat) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]:
```suggestion
   * after making {@code events}'s TimeRanges unavailable for scheduling.
```

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 8]
```diff
@@ -14,10 +14,75 @@
...
+    }
+
+    // Remove TimeRanges that are too short.
+    availableTimes.removeIf(item -> item.duration() < duration);
+
+    return availableTimes;
+  }
+
+  /**
+   * Returns a collection of TimeRanges of TimeRanges in {@code availableTimes} that have been
```
Comment (billie-cat) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]:
```suggestion
   * Returns a collection of TimeRanges in {@code availableTimes} that have been
```

**Review by billie-cat — APPROVED** [review]

**Pull request merged**